### DISCUSSION TITLE
Suggestion to Collections

### DISCUSSION THREAD

**Discussion by lorranpalmeira** [dotnet/csharplang#1162] [General]
`IList<string> MyList = new List<>();`

Instead of

`IList<string> MyList = new List<string>();`

**Comment by yaakov-h** [comment]
#100 probably relevant here.

**Comment by stakx** [comment]
Type expressions such as `List<>` where the generic arguments are missing already have a distinct meaning: these refer to open generic types, or in reflection parlance, to generic type definitions. While overloading these expressions with a new meaning &mdash; generic arguments are inferred &mdash; would save us a modest number of keystrokes here and there, it would also introduce ambiguity into the language, and the potential of misunderstanding such expressions' intent.

```csharp
IList<string> list1 = new List<>();
IList<string> list2 = (IList<>)Activator.CreateInstance(typeof(List<>));
```

A novice C# programmer might then very well wonder why the first line works, while the second does not. Even to a more experienced programmer, it might not be immediately obvious where generic argument inference will happen, and where it won't happen.

In my humble opinion, this kind of ambiguity is unnecessary and should be avoided. It's not worth the few saved keystrokes.

VB.NET's `Nothing` keyword comes to mind. Judging by its surface appearance, i. e. by its name, a novice programmer might easily think that it means "the null reference", only to be surprised by things like `Dim i As Integer = Nothing`. In truth, `Nothing` is much closer in meaning to C# 7.1's `default` than to `null`. (I've spent years both with C# *and* VB.NET, and I still stumble over this.)

I can see the proposed language change introduce the same kind of "deception at first glance".

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/281

**Comment by aluanhaddad** [comment]
@yaakov-h the only issue their is that it doesn't work for abstract types. The example in this post uses an `IList<string>` which wouldn't work by way of #100.

That said, I don't like the proposed syntax of this proposal. This is C# and it should not be ugly. In particular it should not be ugly like Java:smiley:

**Comment by yaakov-h** [comment]
@aluanhaddad that would be why I said it was relevant, not a duplicate. 😊

**Comment by aluanhaddad** [comment]
@yaakov-h true, I read way too much into your pleasantly terse comment 😉
